In [2]:
!pip install librosa

In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.api.types import is_string_dtype
from pandas.api.types import is_numeric_dtype
import librosa
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# 2.1 Getting your data 

We want to read the datasets to try to understand which key we can use to merge them together.

First of all we want to fill empty values in the datasets.
The below function take as an input a dataframe and check if in the columns there are any missing values of numeric or string types and fills this value with an empty string, in case of strings, or with a mean of the values of that column in case of numbers.

In [4]:
def filler(dataframe):
    for col in dataframe.columns:
        if dataframe[col].isnull().any():
            if is_string_dtype(dataframe[col]):
                dataframe[col] = dataframe[col].fillna("")
            elif is_numeric_dtype(dataframe[col]):
                dataframe[col] = dataframe[col].fillna(dataframe[col].mean())

We take the datasets needed:

In [6]:
tracks = pd.read_csv("tracks.csv")
echonest = pd.read_csv("echonest.csv")
features = pd.read_csv("features.csv")

In [7]:
tracks.head()

track_id  album_comments   album_date_created  album_date_released  \
0         2               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
1         3               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
2         5               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
3        10               0  2008-11-26 01:45:08  2008-02-06 00:00:00   
4        20               0  2008-11-26 01:45:05  2009-01-06 00:00:00   

  album_engineer  album_favorites  album_id  \
0            NaN                4         1   
1            NaN                4         1   
2            NaN                4         1   
3            NaN                4         6   
4            NaN                2         4   

                           album_information  album_listens album_producer  \
0                                    <p></p>           6073            NaN   
1                                    <p></p>           6073            NaN   
2                                    <p></p>           6073            NaN   
3                                        NaN          47632            NaN   
4  <p> "spiritual songs" from Nicky Cook</p>           2710            NaN   

   ... track_information track_interest  track_language_code  \
0  ...               NaN           4656                   en   
1  ...               NaN           1470                   en   
2  ...               NaN           1933                   en   
3  ...               NaN          54881                   en   
4  ...               NaN            978                   en   

                                       track_license track_listens  \
0  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1293   
1  Attribution-NonCommercial-ShareAlike 3.0 Inter...           514   
2  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1151   
3  Attribution-NonCommercial-NoDerivatives (aka M...         50135   
4  Attribution-NonCommercial-NoDerivatives (aka M...           361   

  track_lyricist track_number track_publisher  track_tags      track_title  
0            NaN            3             NaN          []             Food  
1            NaN            4             NaN          []     Electric Ave  
2            NaN            6             NaN          []       This World  
3            NaN            1             NaN          []          Freeway  
4            NaN            3             NaN          []  Spiritual Level  

[5 rows x 53 columns]

In [8]:
tracks.shape

(106574, 53)

In [9]:
echonest.head()

track_id  audio_features_acousticness  audio_features_danceability  \
0         2                     0.416675                     0.675894   
1         3                     0.374408                     0.528643   
2         5                     0.043567                     0.745566   
3        10                     0.951670                     0.658179   
4       134                     0.452217                     0.513238   

   audio_features_energy  audio_features_instrumentalness  \
0               0.634476                         0.010628   
1               0.817461                         0.001851   
2               0.701470                         0.000697   
3               0.924525                         0.965427   
4               0.560410                         0.019443   

   audio_features_liveness  audio_features_speechiness  audio_features_tempo  \
0                 0.177647                    0.159310               165.922   
1                 0.105880                    0.461818               126.957   
2                 0.373143                    0.124595               100.260   
3                 0.115474                    0.032985               111.562   
4                 0.096567                    0.525519               114.290   

   audio_features_valence metadata_album_date  ... temporal_features_214  \
0                0.576661                 NaN  ...             -1.992303   
1                0.269240                 NaN  ...             -1.582331   
2                0.621661                 NaN  ...             -2.288358   
3                0.963590          2008-03-11  ...             -3.662988   
4                0.894072                 NaN  ...             -1.452696   

   temporal_features_215 temporal_features_216  temporal_features_217  \
0               6.805694              0.233070               0.192880   
1               8.889308              0.258464               0.220905   
2              11.527109              0.256821               0.237820   
3              21.508228              0.283352               0.267070   
4               2.356398              0.234686               0.199550   

  temporal_features_218 temporal_features_219  temporal_features_220  \
0              0.027455               0.06408                3.67696   
1              0.081368               0.06413                6.08277   
2              0.060122               0.06014                5.92649   
3              0.125704               0.08082                8.41401   
4              0.149332               0.06440               11.26707   

   temporal_features_221  temporal_features_222  temporal_features_223  
0                3.61288              13.316690             262.929749  
1                6.01864              16.673548             325.581085  
2                5.86635              16.013849             356.755737  
3                8.33319              21.317064             483.403809  
4               11.20267              26.454180             751.147705  

[5 rows x 250 columns]

In [10]:
echonest.shape

(13129, 250)

In [11]:
features.head()

track_id  chroma_cens_kurtosis_01  chroma_cens_kurtosis_02  \
0         2                 7.180653                 5.230309   
1         3                 1.888963                 0.760539   
2         5                 0.527563                -0.077654   
3        10                 3.702245                -0.291193   
4        20                -0.193837                -0.198527   

   chroma_cens_kurtosis_03  chroma_cens_kurtosis_04  chroma_cens_kurtosis_05  \
0                 0.249321                 1.347620                 1.482478   
1                 0.345297                 2.295201                 1.654031   
2                -0.279610                 0.685883                 1.937570   
3                 2.196742                -0.234449                 1.367364   
4                 0.201546                 0.258556                 0.775204   

   chroma_cens_kurtosis_06  chroma_cens_kurtosis_07  chroma_cens_kurtosis_08  \
0                 0.531371                 1.481593                 2.691455   
1                 0.067592                 1.366848                 1.054094   
2                 0.880839                -0.923192                -0.927232   
3                 0.998411                 1.770694                 1.604566   
4                 0.084794                -0.289294                -0.816410   

   chroma_cens_kurtosis_09  ...  tonnetz_std_04  tonnetz_std_05  \
0                 0.866868  ...        0.054125        0.012226   
1                 0.108103  ...        0.063831        0.014212   
2                 0.666617  ...        0.040730        0.012691   
3                 0.521217  ...        0.074358        0.017952   
4                 0.043851  ...        0.095003        0.022492   

   tonnetz_std_06  zcr_kurtosis_01  zcr_max_01  zcr_mean_01  zcr_median_01  \
0        0.012111         5.758890    0.459473     0.085629       0.071289   
1        0.017740         2.824694    0.466309     0.084578       0.063965   
2        0.014759         6.808415    0.375000     0.053114       0.041504   
3        0.013921        21.434212    0.452148     0.077515       0.071777   
4        0.021355        16.669037    0.469727     0.047225       0.040039   

   zcr_min_01  zcr_skew_01  zcr_std_01  
0    0.000000     2.089872    0.061448  
1    0.000000     1.716724    0.069330  
2    0.000000     2.193303    0.044861  
3    0.000000     3.542325    0.040800  
4    0.000977     3.189831    0.030993  

[5 rows x 519 columns]

In [12]:
features.shape

(106574, 519)

We can notice that a common column is **track_id**, so let's check if there are any missing values:

In [13]:
tracks["track_id"].isnull().any()

False

In [14]:
echonest["track_id"].isnull().any()

False

In [15]:
features["track_id"].isnull().any()

False

Given that there are no missing values in "**track_id**" columns in any dataset, we choose to use it as the key for the merge of the datasets.

I build a copy and work on it to don't damage the original files.

In [16]:
tracks_c = tracks.copy()
echonest_c = echonest.copy()
features_c = features.copy()

In [17]:
# First we are gonna fill the datasets using the function filler()
filler(tracks_c)
filler(echonest_c)
filler(features_c)

In [18]:
tracks_c.head()

track_id  album_comments   album_date_created  album_date_released  \
0         2               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
1         3               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
2         5               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
3        10               0  2008-11-26 01:45:08  2008-02-06 00:00:00   
4        20               0  2008-11-26 01:45:05  2009-01-06 00:00:00   

  album_engineer  album_favorites  album_id  \
0                               4         1   
1                               4         1   
2                               4         1   
3                               4         6   
4                               2         4   

                           album_information  album_listens album_producer  \
0                                    <p></p>           6073                  
1                                    <p></p>           6073                  
2                                    <p></p>           6073                  
3                                                     47632                  
4  <p> "spiritual songs" from Nicky Cook</p>           2710                  

   ... track_information track_interest  track_language_code  \
0  ...                             4656                   en   
1  ...                             1470                   en   
2  ...                             1933                   en   
3  ...                            54881                   en   
4  ...                              978                   en   

                                       track_license track_listens  \
0  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1293   
1  Attribution-NonCommercial-ShareAlike 3.0 Inter...           514   
2  Attribution-NonCommercial-ShareAlike 3.0 Inter...          1151   
3  Attribution-NonCommercial-NoDerivatives (aka M...         50135   
4  Attribution-NonCommercial-NoDerivatives (aka M...           361   

  track_lyricist track_number track_publisher  track_tags      track_title  
0                           3                          []             Food  
1                           4                          []     Electric Ave  
2                           6                          []       This World  
3                           1                          []          Freeway  
4                           3                          []  Spiritual Level  

[5 rows x 53 columns]

In [19]:
echonest_c.head()

track_id  audio_features_acousticness  audio_features_danceability  \
0         2                     0.416675                     0.675894   
1         3                     0.374408                     0.528643   
2         5                     0.043567                     0.745566   
3        10                     0.951670                     0.658179   
4       134                     0.452217                     0.513238   

   audio_features_energy  audio_features_instrumentalness  \
0               0.634476                         0.010628   
1               0.817461                         0.001851   
2               0.701470                         0.000697   
3               0.924525                         0.965427   
4               0.560410                         0.019443   

   audio_features_liveness  audio_features_speechiness  audio_features_tempo  \
0                 0.177647                    0.159310               165.922   
1                 0.105880                    0.461818               126.957   
2                 0.373143                    0.124595               100.260   
3                 0.115474                    0.032985               111.562   
4                 0.096567                    0.525519               114.290   

   audio_features_valence metadata_album_date  ... temporal_features_214  \
0                0.576661                      ...             -1.992303   
1                0.269240                      ...             -1.582331   
2                0.621661                      ...             -2.288358   
3                0.963590          2008-03-11  ...             -3.662988   
4                0.894072                      ...             -1.452696   

   temporal_features_215 temporal_features_216  temporal_features_217  \
0               6.805694              0.233070               0.192880   
1               8.889308              0.258464               0.220905   
2              11.527109              0.256821               0.237820   
3              21.508228              0.283352               0.267070   
4               2.356398              0.234686               0.199550   

  temporal_features_218 temporal_features_219  temporal_features_220  \
0              0.027455               0.06408                3.67696   
1              0.081368               0.06413                6.08277   
2              0.060122               0.06014                5.92649   
3              0.125704               0.08082                8.41401   
4              0.149332               0.06440               11.26707   

   temporal_features_221  temporal_features_222  temporal_features_223  
0                3.61288              13.316690             262.929749  
1                6.01864              16.673548             325.581085  
2                5.86635              16.013849             356.755737  
3                8.33319              21.317064             483.403809  
4               11.20267              26.454180             751.147705  

[5 rows x 250 columns]

In [20]:
features_c.head()

track_id  chroma_cens_kurtosis_01  chroma_cens_kurtosis_02  \
0         2                 7.180653                 5.230309   
1         3                 1.888963                 0.760539   
2         5                 0.527563                -0.077654   
3        10                 3.702245                -0.291193   
4        20                -0.193837                -0.198527   

   chroma_cens_kurtosis_03  chroma_cens_kurtosis_04  chroma_cens_kurtosis_05  \
0                 0.249321                 1.347620                 1.482478   
1                 0.345297                 2.295201                 1.654031   
2                -0.279610                 0.685883                 1.937570   
3                 2.196742                -0.234449                 1.367364   
4                 0.201546                 0.258556                 0.775204   

   chroma_cens_kurtosis_06  chroma_cens_kurtosis_07  chroma_cens_kurtosis_08  \
0                 0.531371                 1.481593                 2.691455   
1                 0.067592                 1.366848                 1.054094   
2                 0.880839                -0.923192                -0.927232   
3                 0.998411                 1.770694                 1.604566   
4                 0.084794                -0.289294                -0.816410   

   chroma_cens_kurtosis_09  ...  tonnetz_std_04  tonnetz_std_05  \
0                 0.866868  ...        0.054125        0.012226   
1                 0.108103  ...        0.063831        0.014212   
2                 0.666617  ...        0.040730        0.012691   
3                 0.521217  ...        0.074358        0.017952   
4                 0.043851  ...        0.095003        0.022492   

   tonnetz_std_06  zcr_kurtosis_01  zcr_max_01  zcr_mean_01  zcr_median_01  \
0        0.012111         5.758890    0.459473     0.085629       0.071289   
1        0.017740         2.824694    0.466309     0.084578       0.063965   
2        0.014759         6.808415    0.375000     0.053114       0.041504   
3        0.013921        21.434212    0.452148     0.077515       0.071777   
4        0.021355        16.669037    0.469727     0.047225       0.040039   

   zcr_min_01  zcr_skew_01  zcr_std_01  
0    0.000000     2.089872    0.061448  
1    0.000000     1.716724    0.069330  
2    0.000000     2.193303    0.044861  
3    0.000000     3.542325    0.040800  
4    0.000977     3.189831    0.030993  

[5 rows x 519 columns]

Once we got all the datasets well filled, we can start merging them using as key the **track_id** column.

In [21]:
final = tracks_c.merge(features_c.merge(echonest_c, on = "track_id"), on = "track_id")
print(final.shape)
final.head()

(13129, 820)


track_id  album_comments   album_date_created  album_date_released  \
0         2               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
1         3               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
2         5               0  2008-11-26 01:44:45  2009-01-05 00:00:00   
3        10               0  2008-11-26 01:45:08  2008-02-06 00:00:00   
4       134               0  2008-11-26 01:44:45  2009-01-05 00:00:00   

  album_engineer  album_favorites  album_id album_information  album_listens  \
0                               4         1           <p></p>           6073   
1                               4         1           <p></p>           6073   
2                               4         1           <p></p>           6073   
3                               4         6                            47632   
4                               4         1           <p></p>           6073   

  album_producer  ... temporal_features_214 temporal_features_215  \
0                 ...             -1.992303              6.805694   
1                 ...             -1.582331              8.889308   
2                 ...             -2.288358             11.527109   
3                 ...             -3.662988             21.508228   
4                 ...             -1.452696              2.356398   

   temporal_features_216 temporal_features_217 temporal_features_218  \
0               0.233070              0.192880              0.027455   
1               0.258464              0.220905              0.081368   
2               0.256821              0.237820              0.060122   
3               0.283352              0.267070              0.125704   
4               0.234686              0.199550              0.149332   

  temporal_features_219 temporal_features_220 temporal_features_221  \
0               0.06408               3.67696               3.61288   
1               0.06413               6.08277               6.01864   
2               0.06014               5.92649               5.86635   
3               0.08082               8.41401               8.33319   
4               0.06440              11.26707              11.20267   

   temporal_features_222 temporal_features_223  
0              13.316690            262.929749  
1              16.673548            325.581085  
2              16.013849            356.755737  
3              21.317064            483.403809  
4              26.454180            751.147705  

[5 rows x 820 columns]

As we can see, the number of rows is almost 13k, as demanded.

## 2.2 Choose your features (variables)!

We notice that the number of features is very big (820 columns!) and we want to reduce the number of variables we have to work with.

So let's apply dimensionality reduction!

### Principal Component Analysis

Principal component analysis, is a statistical technique to convert high dimensional data to low dimensional data by selecting the most important features that capture maximum information about the dataset. The features are selected on the basis of variance that they cause in the output. The feature that causes highest variance is the first principal component. The feature that is responsible for second highest variance is considered the second principal component, and so on. It is important to mention that principal components do not have any correlation with each other.

#### Standardizing the Numeric Features

Since PCA is effected by scale, we need to scale the features in our data before applying PCA. We used StandardScaler to standardize the dataset’s features onto unit scale (mean = 0 and variance = 1) which is a requirement for the optimal performance of algorithms. 

In [24]:
# First we need to chose numeric columns 
num_cols = [col for col in final.columns if final[col].dtypes != "O"]
num_cols

['track_id',
 'album_comments',
 'album_favorites',
 'album_id',
 'album_listens',
 'album_tracks',
 'artist_comments',
 'artist_favorites',
 'artist_id',
 'artist_latitude',
 'artist_longitude',
 'track_bit_rate',
 'track_comments',
 'track_duration',
 'track_favorites',
 'track_interest',
 'track_listens',
 'track_number',
 'chroma_cens_kurtosis_01',
 'chroma_cens_kurtosis_02',
 'chroma_cens_kurtosis_03',
 'chroma_cens_kurtosis_04',
 'chroma_cens_kurtosis_05',
 'chroma_cens_kurtosis_06',
 'chroma_cens_kurtosis_07',
 'chroma_cens_kurtosis_08',
 'chroma_cens_kurtosis_09',
 'chroma_cens_kurtosis_10',
 'chroma_cens_kurtosis_11',
 'chroma_cens_kurtosis_12',
 'chroma_cens_max_01',
 'chroma_cens_max_02',
 'chroma_cens_max_03',
 'chroma_cens_max_04',
 'chroma_cens_max_05',
 'chroma_cens_max_06',
 'chroma_cens_max_07',
 'chroma_cens_max_08',
 'chroma_cens_max_09',
 'chroma_cens_max_10',
 'chroma_cens_max_11',
 'chroma_cens_max_12',
 'chroma_cens_mean_01',
 'chroma_cens_mean_02',
 'chroma_cens

In [29]:
# We changed the types of numeric columns as float to apply Standard Scaler
final[num_cols] = final[num_cols].astype(float)
final.dtypes

track_id                 float64
album_comments           float64
album_date_created        object
album_date_released       object
album_engineer            object
                          ...   
temporal_features_219    float64
temporal_features_220    float64
temporal_features_221    float64
temporal_features_222    float64
temporal_features_223    float64
Length: 820, dtype: object

In [40]:
# Separating out the non-numeric features
df_1 = final.drop(columns=num_cols)
df_1.dtypes

album_date_created          object
album_date_released         object
album_engineer              object
album_information           object
album_producer              object
album_tags                  object
album_title                 object
album_type                  object
artist_active_year_begin    object
artist_active_year_end      object
artist_associated_labels    object
artist_bio                  object
artist_date_created         object
artist_location             object
artist_members              object
artist_name                 object
artist_related_projects     object
artist_tags                 object
artist_website              object
artist_wikipedia_page       object
set_split                   object
set_subset                  object
track_composer              object
track_date_created          object
track_date_recorded         object
track_genre_top             object
track_genres                object
track_genres_all            object
track_information   

In [31]:
# Separating out the numeric features
x = final.loc[:, num_cols].values

In [32]:
# Standardizing the features
x = StandardScaler().fit_transform(x)

In [48]:
# Initialize the PCA class by passing the number of components to the constructor.
pca = PCA(n_components=5)
# Call the fit and then transform methods by passing the feature set to these methods. 
# The transform method returns the specified number of principal components.
principalComponents = pca.fit_transform(x)
# Creating Principal data frame
principalDf = pd.DataFrame(data = principalComponents
             , columns = ['principal component 1', 'principal component 2','principal component 3','principal component 4','principal component 5'])

In [49]:
print('Explained variation per principal component: {}'.format(pca.explained_variance_ratio_))

Explained variation per principal component: [0.10472581 0.07410088 0.0353129  0.02836919 0.02501076]


In [50]:
finalDf = pd.concat([principalDf, df_1], axis = 1)

In [52]:
finalDf.head(10)

principal component 1  principal component 2  principal component 3  \
0               3.074543              -2.769912              -5.858218   
1               5.774790              -5.364765              -5.397493   
2               2.652435              -4.121103              -4.686471   
3               9.135772               3.874803              -5.557803   
4               1.307063              -9.400518              -4.458017   
5               4.063851               2.598676               1.049471   
6              -6.592735              -3.592278               4.241709   
7             -12.182820               2.051396               1.524762   
8              -4.686128              -1.072051              -3.854670   
9               8.312074               3.402502              -5.405963   

   principal component 4  principal component 5   album_date_created  \
0               4.197439               3.259071  2008-11-26 01:44:45   
1              -1.474541              -3.646719  2008-11-26 01:44:45   
2               0.059510              -2.877743  2008-11-26 01:44:45   
3               0.749018              -1.304397  2008-11-26 01:45:08   
4               0.951766              -4.015965  2008-11-26 01:44:45   
5               1.678790               3.190033  2008-11-26 01:49:57   
6              -1.613454               0.856779  2008-11-26 01:49:59   
7              -3.027803              -1.933654  2008-11-26 01:49:57   
8              -6.691118              -3.151397  2008-11-26 01:50:03   
9              -1.735372               1.878881  2008-11-26 01:50:07   

   album_date_released album_engineer  \
0  2009-01-05 00:00:00                  
1  2009-01-05 00:00:00                  
2  2009-01-05 00:00:00                  
3  2008-02-06 00:00:00                  
4  2009-01-05 00:00:00                  
5  2009-01-16 00:00:00                  
6  2007-05-22 00:00:00                  
7  2009-01-16 00:00:00                  
8  2005-01-25 00:00:00                  
9  2009-01-06 00:00:00                  

                                   album_information  \
0                                            <p></p>   
1                                            <p></p>   
2                                            <p></p>   
3                                                      
4                                            <p></p>   
5  <p>A full ensamble of strings, drums, electron...   
6  <p>Alec K. Redfearn &amp; The Eyesores: Ellen ...   
7  <p>A full ensamble of strings, drums, electron...   
8  <p>Recorded at Sound Station Seven and at home...   
9  <p><em>A</em>lthough recorded in Providence in...   

                   album_producer  ...  \
0                                  ...   
1                                  ...   
2                                  ...   
3                                  ...   
4                                  ...   
5                                  ...   
6  Alec K. Refearn, Rob Pemberton  ...   
7                                  ...   
8                                  ...   
9                    Tom Buckland  ...   

                                       track_license track_lyricist  \
0  Attribution-NonCommercial-ShareAlike 3.0 Inter...                  
1  Attribution-NonCommercial-ShareAlike 3.0 Inter...                  
2  Attribution-NonCommercial-ShareAlike 3.0 Inter...                  
3  Attribution-NonCommercial-NoDerivatives (aka M...                  
4  Attribution-NonCommercial-ShareAlike 3.0 Inter...                  
5  Attribution-Noncommercial-No Derivative Works ...                  
6  Attribution-Noncommercial-No Derivative Works ...                  
7  Attribution-Noncommercial-No Derivative Works ...                  
8  Attribution-Noncommercial-No Derivative Works ...                  
9  Attribution-Noncommercial-No Derivative Works ...                  

  track_publisher track_tags           track_title metadata_album_date  \
0      